SECP256k1 support

kozross opened the pull request:
This closes #4233, together with [the relevant `cardano-base` PR](https://github.com/input-output-hk/cardano-base/pull/252). <s>Currently, this PR depends on the MLabs fork of `cardano-base`; once the linked PR merges, I will amend to work against `master` again. This shouldn't affect review, as while this depends on functionality from `cardano-base`, nothing in here is affected by precisely what branch of `cardano-base` we pull from.</s>

**Edit:** This is now done.

I've added some property-based tests using Hedgehog to ensure all possible paths are equally covered. Specifically, we check each of the following:

* Malformed pubkey
* Malformed signature
* Malformed message hash
* Use of the wrong verification key
* Checking the wrong signature for a given message
* The happy path (that is, everything works)

As per @kwxm's feedback on [my prior work on `plutus-tx`](https://github.com/input-output-hk/plutus/pull/4322#issuecomment-1012052868), I've raised the test count high enough to ensure that fewer than 1 in a 1000 executions fail due to probabilistic coverage failures (a flaw in Hedgehog, and one we can't avoid).

A few concerns remain:

* The builtin for SECP256k1 verification is _extremely_ demanding of how its inputs are formatted, but currently, this is left entirely to the user to ensure: all arguments are `BuiltinByteString`s. I didn't know if we wanted helpers (in `plutus-tx` or elsewhere) to make this process less nasty with `newtype` wrappers and helper functionality, so I haven't added it.
* The tests for `untyped-plutus-core` are far too few in number: almost everything in there _should_ be a property test, but even those which are run 200 tests, which is too few to indicate anything. For a demonstration of why: I found multiple bugs in how I implemented the SECP256k1 builtin only after over a _thousand_ tests, and even then it took several attempts! Furthermore, coverage is not checked _at all_, which is alarming. The tests I wrote for the SECP256k1 primitive are vulnerable to neither.
* As per @michaelpj's suggestion [here](https://github.com/input-output-hk/plutus/issues/4233#issuecomment-1021086335), the costing function is currently `mempty`. I don't know how much of an issue this is.

ysangkok commented:
It's confusing to refer to all of this as "SECP256k1 support" since SECP256k1 is just a curve. Since you are using secp256k1-haskell, it seems you do not support the Schnorr ([BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki)) signatures recently added to Bitcoin. So I suspect that you have in fact added ECDSA signature support using the SECP256k1 curve. Is that correct?

effectfully commented on plutus-core/plutus-core.cabal:
```diff
@@ -435,11 +436,13 @@ test-suite untyped-plutus-core-test
     type: exitcode-stdio-1.0
     main-is: Spec.hs
     hs-source-dirs: untyped-plutus-core/test
+    ghc-options: -O2 -threaded -rtsopts -with-rtsopts=-N
```
It's funny we didn't have that already.

kozross replied:
I actually needed to add this to allow me to ensure that Hedgehog coverage didn't nondeterministically fail too often in a reasonable amount of time. But yes, it's strange that this isn't being used in _all_ test suites.

effectfully commented on plutus-core/plutus-core/src/Crypto.hs:
```diff
@@ -17,3 +28,25 @@ verifySignature pubKey msg sig =
             <*> pure msg
             <*> signature sig
+
+verifySECP256k1Signature
+  :: BS.ByteString -- ^ Public key
+  -> BS.ByteString -- ^ Signature
+  -> BS.ByteString -- ^ Message hash
+  -> Emitter (EvaluationResult Bool)
+verifySECP256k1Signature pk sig msg =
+  case DSIGN.rawDeserialiseVerKeyDSIGN @SECP256k1DSIGN pk of
+    Nothing -> do
+      emitM "SECP256k1: Given invalid signing key."
+      pure EvaluationFailure
+    Just pk' -> case DSIGN.rawDeserialiseSigDSIGN @SECP256k1DSIGN sig of
+      Nothing -> do
+        emitM "SECP256k1: Given invalid signature."
+        pure EvaluationFailure
+      Just sig' -> case SECP.msg msg of
+        Nothing -> do
+          emitM "SECP256k1: Given invalid message hash."
+          pure EvaluationFailure
+        Just msg' -> case DSIGN.verifyDSIGN () pk' msg' sig' of
+          Left _   -> pure . pure $ False
+          Right () -> pure . pure $ True
```
@kwxm bad news for ya, we have another emitting bulitin.

kozross replied:
Is this a problem in some sense I'm not aware of?

effectfully replied:
@kozross, not a concern for you, just a pain in the ass to describe in the specification.

kwxm replied:
I think we were really expecting this to be used only in the in the implementation of `trace` and not in arbitrary builtins, but I suppose it's harmless and it's commendable to give programmers extra feedback with something that it would be easy to get wrong.

I think from the point of specification, I'd inclined to just not mention this in the specification on the basis that the error messages are a detail specific to this particular implementation.  Trying to include the details of the logging mechanism in the specification would complicate things quite a bit, so maybe we could say that an implementation of Plutus Core may choose to provide facilities for logging errors as long as they don't have any effect on the semantics of built-in functions;  that's not really got anything to do with this PR though.

Off-topic, this makes me wonder if the logging mechanism could be used to do unexpected things in contract code. Could the off-chain part of a contract run some validation code (also off the chain), look at the log files and then branch on the contents to determine the future behaviour of the contract (which you could do even using `trace`)? That might provide a way to leak more information from validation than the `true`/`false` answer that we expect, but would that actually be a problem?

effectfully commented on plutus-tx-plugin/plutus-tx-plugin.cabal:
```diff
@@ -130,6 +130,7 @@ test-suite plutus-tx-tests
     -- NOTE: -g makes the plugin give better errors
     ghc-options: -g
 
+
```
Random change in an otherwise untouched file.

effectfully commented on plutus-tx/src/PlutusTx/Builtins/Internal.hs:
```diff
@@ -197,6 +199,17 @@ verifySignature :: BuiltinByteString -> BuiltinByteString -> BuiltinByteString -
+{-# NOINLINE verifySECP256k1Signature #-}
+verifySECP256k1Signature :: BuiltinByteString -> BuiltinByteString -> BuiltinByteString -> BuiltinBool
+verifySECP256k1Signature (BuiltinByteString pk) (BuiltinByteString sig) (BuiltinByteString msg) =
+  case Crypto.verifySECP256k1Signature pk sig msg of
+    Emitter f -> case runIdentity (f go) of
+      EvaluationFailure   -> mustBeReplaced "SECP256k1 verification errored."
+      EvaluationSuccess b -> BuiltinBool b
+  where
+    go :: Text -> Identity ()
+    go t = Identity . trace (BuiltinString t) $ ()
```
Have you tried running it to see if it indeed traces something?

The monad instance for `Identity` is defined as

```haskell
instance Monad Identity where
    m >>= k  = k (runIdentity m)
```

i.e. impure effects in `m` are not performed unless `k` forces its arguments, which we may or may not do.

What if you just made `go` run in `EvaluationResult` rather than `Identity`? Something like

```haskell
go :: Text -> EvaluationResult ()
go t = trace (BuiltinString t) $ EvaluationSuccess ()
``` 

(note that `trace` is outside of `EvaluationSuccess` rather than inside, because `EvaluationResult` is not strict and so similar considerations apply)

kozross replied:
I wasn't sure about the right way to do this, so if you think it should be using `trace` like that, I'll change it.

effectfully reviewed (approved):
The Haskell part looks great to me, apart from `PlutusTx.Builtins.Internal.verifySECP256k1Signature`, which I think could be better.

Can't comment on the crypto part.

effectfully commented:
Ah, yeah, it must conflict horribly with my recent refactoring, I'll fix it for you.

effectfully commented:
> Ah, yeah, it must conflict horribly with my recent refactoring, I'll fix it for you.

[Done](https://github.com/mlabs-haskell/plutus/pull/2) (not horribly at all).

michaelpj commented:
For some reason it won't let me respond to the existing comments...

I agree that it's harmless to include tracing here. It's interesting that this is the first builtin other than `trace` that emits things, but probably good.

I'm not worried about specifying the tracing. We deliberately don't specify anything to do with log emitting since it's not observable in the main behaviour of the program. 

> Could the off-chain part of a contract run some validation code (also off the chain), look at the log files and then branch on the contents to determine the future behaviour of the contract (which you could do even using trace)? 

Yes, but you can do whatever you want off-chain. You can run your PLC through a symbolic executor and buy a sandwich iff it returns true on Tuesday 22nd 2024. So I don't think we need to care about that.

kwxm commented:
> I agree that it's harmless to include tracing here. It's interesting that this is the first builtin other than `trace` that emits things, but probably good.

I suppose there's not much that can go wrong with most of the existing builtins (mostly out of range errors), so there's no need give details when something goes wrong.  Maybe we could be more informative with some of the other ones though.  I 'll have to benchmark this for costing purposes at some point and while I'm doing that I'll check that the logging isn't adding any overhead.  It's highly unlikely that it will be since nothing at all should be happening on chain, but there's no harm checking.

kwxm commented on plutus-core/plutus-core/src/Crypto.hs:
```diff
@@ -17,3 +28,25 @@ verifySignature pubKey msg sig =
+          emitM "SECP256k1: Given invalid message hash."
+          pure EvaluationFailure
+        Just msg' -> case DSIGN.verifyDSIGN () pk' msg' sig' of
+          Left _   -> pure . pure $ False
```
I don't understand all the stuff in `DSIGN`, but I see that the `_` is a string.  Is that something that would be worth logging in the `Left` case, or is it stuff that should be caught by the earlier checks anyway?

kwxm reviewed (approved):
Looks good.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -786,6 +791,7 @@ instance Flat DefaultFun where
               Sha3_256                 -> 19
               Blake2b_256              -> 20
               VerifySignature          -> 21
+              VerifySECP256k1Signature -> 51
```
In fact, maybe that should go at the end.  If someone else wants to add a new builtin they might look at the end of the list and think that 51's the next free tag.

kozross replied:
So do you suggest I use the largest number available for `VerifySECP256k1Signature` instead of `51`? There's no specified system of numbering these in the documentation of the module.

kwxm replied:
No, I meant that if you look at the end of that list of cases it says
```
              UnListData               -> 44
              UnIData                  -> 45
              UnBData                  -> 46
              EqualsData               -> 47
              MkPairData               -> 48
              MkNilData                -> 49
              MkNilPairData            -> 50
```
so it would be easy to think that the next free number is 51.  Of course if we arrange them numerically then it moves `VerifySECP256k1Signature` away from the rest of the cryptographic builtins, so you get a different kind of confusion.

kwxm commented on plutus-tx/src/PlutusTx/Builtins.hs:
```diff
@@ -157,6 +158,24 @@ greaterThanEqualsByteString x y = BI.ifThenElse (BI.lessThanByteString x y) Fals
+  -> Bool
+verifySECP256k1Signature pk sig msg =
+  fromBuiltin (BI.verifySECP256k1Signature pk sig msg)
+
```
Some digging through the repositories (with @michaelpj 's help) reveals that there is a [DSIGN interface](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs) for `Ed255911 as well. We should probably switch to using that for `verifySignature` instead, but unfortunately we won't be able to rearrange the arguments for reasons of backward compatibility.  Anyway, the order above seems the right one for things using `DSIGN` and we should probably use the same order for any further signature checking functions we implement.

kozross replied:
I agree that being consistent in _some_ sense is better here, but this is a concern far outside this PR. I picked the argument order the way I did for precisely the reason you said: by matching the underlying library function, I minimize the amount of mistakes I would make by forgetting which argument goes where.

kwxm replied:
> precisely the reason you said: by matching the underlying library function

I think I was wrong about that: see the comment in Crypto.hs.  I'm thoroughly confused now though.

kozross commented on plutus-core/untyped-plutus-core/test/Evaluation/Builtins/SECP256k1.hs:
```diff
@@ -0,0 +1,282 @@
+      SECP.Msg ->
+      Gen SECP256k1NoErrorCase
+    mkWrongKey sk pk msg = do
+      pkBad <- Gen.filter (/= pk) genPubKey
```
It's very unlikely. Assuming the generator is 'fair' (that is, any given key is equally likely to be generated), the probability that this filter would be tripped is 1 over (2^8)^32, due to pubkey lengths.

kozross commented on plutus-tx/src/PlutusTx/Builtins.hs:
```diff
@@ -157,6 +158,24 @@ greaterThanEqualsByteString x y = BI.ifThenElse (BI.lessThanByteString x y) Fals
+{-# INLINEABLE verifySECP256k1Signature #-}
+-- | Given a SECP256k1 public key, a SECP256k1 signature, and a SECP256k1
+-- message hash (all as 'BuiltinByteString's), verify the hash with that key and
+-- signature.
```
Basically, this is a concern outside of this builtin, `cardano-base` and even `secp256k1-haskell`! Specifically, `secp256k1-haskell` assumes you can somehow obtain this hash, and that's as far as its concerns go.

kozross commented:
@ysangkok Thank you for bringing this to our attention. We do indeed need Schnorr signature support; therefore, I'm marking this PR as WIP for now. I'll keep it current with `plutus`, but to add those will require me to do additional PRs to `cardano-base` at least, and possibly `secp256k1-haskell` as well.

@michaelpj and @kwxm: I have attempted, based on [these instructions here](https://github.com/input-output-hk/plutus/blob/master/plutus-core/cost-model/CostModelGeneration.md) and [your feedback here](https://github.com/input-output-hk/plutus/issues/4375#issuecomment-1029864939), to generate a cost model for `VerifySECP256k1Signature`. However, after running `cabal bench plutus-core:update-cost-model`, the `builtinCostModel.json` doesn't seem to contain any reference to `VerifySECP256k1Signature` at all. I'm pretty sure I missed something, but I don't know what. The cost model in this case is fairly simple: the arguments are all fixed-width, so it should just be a constant function: I only don't know _what_ said constant should be.

kwxm commented:
> I have attempted, based on [these instructions here](https://github.com/input-output-hk/plutus/blob/master/plutus-core/cost-model/CostModelGeneration.md) and https://github.com/input-output-hk/plutus/issues/4375#issuecomment-1029864939, to generate a cost model for VerifySECP256k1Signature.

@kozross Please undo this!!!!  As mentioned in the second link above, these benchmarks MUST be run on a specific machine which at the moment only we have access to (unfortunately, but that's hard to change and there may even be legal difficulties because the machine is provided to us by an external party and we may be limited in who's allowed to access it).  The changes to `costModel.json` involve drastic changes to the costing functions for _all_ of the builtins and it would be disastrous to merge this because it would lead to significant changes in the costs of scripts.  We need the costs to remain relatively consistent and to ensure this we have to use the same machine every time. 

We can generate a suitable cost model once this PR is merged.  In the meantime `mempty` will be OK, especially because the function should have constant cost.  

I did send an email about this to the address on your GitHub account a few days ago, but I'm not sure if you got it.

kozross commented:
@kwxm OK, I'll revert the change. I did receive your email, but the 'reference machine' part must have eluded me.

kwxm commented on plutus-core/plutus-core/src/Crypto.hs:
```diff
@@ -17,3 +28,25 @@ verifySignature pubKey msg sig =
+        Nothing -> do
+          emitM "SECP256k1: Given invalid message hash."
+          pure EvaluationFailure
+        Just msg' -> case DSIGN.verifyDSIGN () pk' msg' sig' of
```
Oh wait!  `verifySECP256k1Signature` takes its arguments in the order pk/sig/msg, but `DSIGN.verifyDSIGN` takes them in the order pk'/msg'/sig': see also the source of the instance [here](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs#L145).  Earlier I thought both of these took their arguments in the same order.  Would it be reasonable to change the order of the arguments in the main function?  That would make it consistent with the order expected by `DSIGN` (and also with the existing `verifySignature` function), which would be helpful if we ever implement other signature schemes covered by `DSIGN`.

kwxm reviewed (commented):
(Sorry, I was trying to leave some comments and I seem to have started another review.)

kwxm commented:
> @kwxm OK, I'll revert the change. I did receive your email, but the 'reference machine' part must have eluded me.

Thanks!  I'm afraid that this situation's not very satisfactory, but it's not clear what else we can do at the moment.  Eventually there'll have to be some community process for this kind of thing, but that's probably quite a long way off at the moment.

michaelpj commented:
I'm glad we discovered the confusion about the signature algorithm before we merged this!

@kozross I'll let you work out what it is that you actually need, I guess the only thing to bear in mind for coming back to this PR is that we should indeed include the algorithm name as well as the curve in the name of the builtin, so we don't get confused in future.

kozross commented:
@michaelpj, @effectfully and @kwxm - I have now added the missing builtin we needed, addressed all concerns (as well as some naming concerns raised both here and in `cardano-base`). This currently depends on the MLabs fork of `cardano-base`, pending [this PR landing](https://github.com/input-output-hk/cardano-base/pull/258).

effectfully commented on plutus-core/plutus-core/src/Crypto.hs:
```diff
@@ -17,3 +34,58 @@ verifySignature pubKey msg sig =
+--
+-- This takes a message /hash/, rather than a general blob of bytes; thus, it is
+-- limited in length.
+verifyEcdsaSecp256k1Signature
+  :: BS.ByteString -- ^ Public key (64 bytes)
+  -> BS.ByteString -- ^ Message hash (32 bytes)
+  -> BS.ByteString -- ^ Signature (64 bytes)
+  -> Emitter (EvaluationResult Bool)
+verifyEcdsaSecp256k1Signature pk msg sig =
+  case DSIGN.rawDeserialiseVerKeyDSIGN @EcdsaSecp256k1DSIGN pk of
+    Nothing -> failWithMessage loc "Invalid verification key."
+    Just pk' -> case DSIGN.rawDeserialiseSigDSIGN @EcdsaSecp256k1DSIGN sig of
```
You use here the definitions that we provide, but evidently what we provide is not suitable for the task. It would be reasonable to expect `failWithMessage loc str *> b` to fail with `EvaluationFailure` without evaluating `b`, but it will evaluate `b`. The problem is with our abstractions, not your code, so my recommendation would be to add a short note saying that existing abstractions are not sufficient and the code needs to be improved and then someone (probably from the Plutus team) would go and actually fix the code with a monad transformer or something. I don't think that should block this PR, since the code looks perfectly fine, it's only the abstractions that are off.

kozross replied:
This is slightly orthogonal to my work though - I only used `failWithMessage` as a refactor for a common 'terminating procedure', never expecting it to be 'sequenced' with anything else. I'm happy to leave such a note, but where would it go? `failWithMessage` isn't the only way to exhibit this issue, and if I put it here, future readers aren't likely to find it.

effectfully replied:
> This is slightly orthogonal to my work though

Absolutely.

> never expecting it to be 'sequenced' with anything else.

Yes, I understand that, but a bad abstraction is a bad abstraction and we should fix it. You're doing the best that you can here with the tools you're given, I do not suggest you should change anything.

> I'm happy to leave such a note, but where would it go? `failWithMessage` isn't the only way to exhibit this issue, and if I put it here, future readers aren't likely to find it.

It's not for future readers, it's for us to keep us annoyed, so that we're motivated enough to fix the issue. Any place would work, so `failWithMessage` is good enough.

kwxm commented on plutus-core/untyped-plutus-core/test/Spec.hs:
```diff
@@ -22,7 +24,9 @@ import UntypedPlutusCore qualified as UPLC
 main :: IO ()
-main = defaultMain $ testGroup "Untyped Plutus Core"
+main = do
+  setLocaleEncoding utf8
```
What's this for?  I'm sure there's a good reason for it and I'm wondering if we should be doing the same elsewhere.

kozross replied:
Without this, you get crashes due to mismatched encodings in (at least my) `nix-shell`.

L-as replied:
To elaborate further: It is assumed that the locale is UTF-8 without setting it to UTF-8 explicitly, meaning that `base` infers the locale from the environment, and this often fails in the context of Nix due to `LANG`, etc. possibly not being set IIRC. This is good practice in general really and should just be the default IMO.

kwxm replied:
I see what you mean: if you set `LC_ALL` to something nonsensical then our current tests start failing with messages like `plutus-core-test: <stdout>: commitBuffer: invalid argument (invalid character)`.  That's useful to know. Thanks!

kwxm commented on plutus-core/untyped-plutus-core/test/Evaluation/Builtins/SECP256k1.hs:
```diff
@@ -0,0 +1,368 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
Very thorough tests!  We should probably steal your ideas and re-use them for the existing `verifySignature` function.

kozross replied:
I aim to be thorough - this isn't the kind of thing you want to be careless with.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -589,6 +591,14 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
+    toBuiltinMeaning VerifyEcdsaSecp256k1Signature =
+        makeBuiltinMeaning
+            verifyEcdsaSecp256k1Signature
+            mempty
```
We still don't have any way for contributors to add cot models, so I'll add models for these when this is merged.

kwxm commented on plutus-core/plutus-core/src/Crypto.hs:
```diff
@@ -17,3 +34,64 @@ verifySignature pubKey msg sig =
+-- This takes a message /hash/, rather than a general blob of bytes; thus, it is
+-- limited in length.
+verifyEcdsaSecp256k1Signature
+  :: BS.ByteString -- ^ Public key (64 bytes)
```
I've added [similar comments](https://github.com/input-output-hk/plutus/pull/4437) about input lengths to the existing `verifySignature` function, and we'll update it to use the `DSIGN` interface at some point as well.  We should probably also give it a more informative name (or at least an alternative) if we can do that without breaking anything.

kozross replied:
I believe @michaelpj said that it'd be too breaking to change the name somewhere else.

kwxm reviewed (approved):
This all looks good, although I haven't dug in to the underlying crypto stuff (yet).

kozross commented:
Windows builds seem to be failing according to the logs.

kozross commented:
@michaelpj I have no access to a Windows machine to test this on, and frankly don't even know how this combination of Nix and Windows works. I'm happy to do something if instructed how, but it's _very_ far out of my understanding.

michaelpj commented:
I can take a look. It would really help if you checked the box to allow commits to your branch by maintainers.

kozross commented:
@michaelpj I would like to, but I'm having trouble finding it. This is all I can see:

![This is all I see.](https://user-images.githubusercontent.com/8463904/157304705-38bc47b3-b6c4-4b7d-b497-7604d2a2c7bf.png)

michaelpj commented:
Hmm, I don't know why that isn't there, normally you can do that :thinking: 

Summon @hamishmack , who set up the windows stuff, perhaps you can give us a hand?

kozross commented:
@michaelpj I think I fixed that issue, but now Agda's blowing up instead.

kozross commented:
@michaelpj In that case, @hamishmack might know why. I took my best guess at the problem. I'll merge your PR in the meantime.

hamishmack commented:
Try this https://github.com/hamishmack/plutus/commit/57469864592d77eded0b0f3983f368d15525d329

bezirg commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -1115,6 +1127,8 @@ instance Flat DefaultFun where
               go 48 = pure MkPairData
               go 49 = pure MkNilData
               go 50 = pure MkNilPairData
+              go 51 = pure VerifyEcdsaSecp256k1Signature
+              go 52 = pure VerifySchnorrSecp256k1Signature
```
```suggestion
              go 52 = pure VerifyEcdsaSecp256k1Signature
              go 53 = pure VerifySchnorrSecp256k1Signature
```

kozross commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -1001,67 +1011,69 @@ decodeBuiltin = dBEBits8 builtinTagWidth
+              Sha3_256                        -> 19
+              Blake2b_256                     -> 20
+              VerifySignature                 -> 21
+              VerifyEcdsaSecp256k1Signature   -> 51
+              VerifySchnorrSecp256k1Signature -> 52
```
Did I get the numbering wrong or did we introduce a new builtin?

bezirg replied:
Hi Koz, 1 new builtin just got merged to master so i thought to let you know. https://github.com/input-output-hk/plutus/commit/0ed57009373635b14e925ed64babe4331187dd44

bezirg commented on plutus-tx/plutus-tx.cabal:
```diff
@@ -1,156 +1,167 @@
+    , cborg
+    , hedgehog
+    , hedgehog-fn
+    , plutus-core     ^>=1.0
```
```suggestion
    , plutus-core:plutus-core-testlib     ^>=1.0
```

michaelpj commented:
> Edit: It seems they were moved into plutus-core-testlib, which means I now can't depend on them.

You can, just see Nikos' suggestion.

michaelpj commented:
Now the build issues are sorted out I think we can return to this. I'm not totally clear where we are on it: I think we're basically good to go? The failing builds are darwin (we have build machine issues, so it's disabled on master); merging master should fix that. And there are a few conflicts.

kozross commented:
@michaelpj I've resolved the conflicts. Other than this, I believe the work is done: I have addressed reviewer concerns, I'm not aiming at any branches (such as for `cardano-base`), and the tests are present and passing.

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:
```diff
@@ -282,6 +284,8 @@ defineBuiltinTerms = do
     defineBuiltinTerm 'Builtins.emptyByteString $ PIR.mkConstant () BS.empty
     defineBuiltinTerm 'Builtins.decodeUtf8 $ mkBuiltin PLC.DecodeUtf8
+    defineBuiltinTerm 'Builtins.verifyEcdsaSecp256k1Signature $ mkBuiltin PLC.VerifyEcdsaSecp256k1Signature
+    defineBuiltinTerm 'Builtins.verifySchnorrSecp256k1Signature $ mkBuiltin PLC.VerifyEcdsaSecp256k1Signature
```
```suggestion
    defineBuiltinTerm 'Builtins.verifySchnorrSecp256k1Signature $ mkBuiltin PLC.VerifySchnorrSecp256k1Signature
```

kozross replied:
I've fixed that (as well as the attendant golden test), but the issue with the parser roundtripping still happens.

michaelpj commented:
The error message is hard to read because megaparsec errors don't `Show` well, but the issue is that you need to add cases for the new builtins to https://github.com/input-output-hk/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/Parser/Builtin.hs.

(@thealmarty can we make this list auto-generated so this doesn't happen in future? I think we should just be able to enumerate `DefaultFun` and populate the list with the value and its pretty-printed version.)

michaelpj commented:
https://github.com/mlabs-haskell/plutus/pull/5

effectfully commented:
> (@thealmarty can we make this list auto-generated so this doesn't happen in future? I think we should just be able to enumerate DefaultFun and populate the list with the value and its pretty-printed version.)

(It's how it used to be. Not only doing this automatically is way more convenient, but also allows us to handle sets of built-in functions other than `DefaultFun`)

kozross commented:
@michaelpj Yeah, I believe it used to work this way, and must have changed without my noticing.

kozross commented:
I can't seem to reproduce this failure locally:

```
      deconstructorData2:             FAIL
        Test output was different from 'test/Plugin/Primitives/deconstructorData2.plc.golden'. Output of ["diff","-u","test/Plugin/Primitives/deconstructorData2.plc.golden","/build/deconstructorData2.plc13-54.actual"]:
        --- test/Plugin/Primitives/deconstructorData2.plc.golden        1970-01-01 00:00:01.000000000 +0000
        +++ /build/deconstructorData2.plc13-54.actual   2022-04-21 19:51:44.217039864 +0000
        @@ -239,6 +239,31 @@
                   (termbind
                     (strict)
                     (vardecl
        +              fst (all a (type) (all b (type) (fun [ [ (con pair) a ] b ] a)))
        +            )
        +            (builtin fstPair)
        +          )
        +          (termbind
        +            (strict)
        +            (vardecl
        +              snd (all a (type) (all b (type) (fun [ [ (con pair) a ] b ] b)))
        +            )
        +            (builtin sndPair)
        +          )
        +          (termbind
        +            (strict)
        +            (vardecl
        +              unsafeDataAsConstr
        +              (fun
        +                (con data)
        +                [ [ (con pair) (con integer) ] [ (con list) (con data) ] ]
        +              )
        +            )
        +            (builtin unConstrData)
        +          )
        +          (termbind
        +            (strict)
        +            (vardecl
                       chooseList
         <truncated>
        Use --accept or increase --size-cutoff to see full output.
        Use -p '/deconstructorData2/' to rerun this test only.
```

I had to repair a similar issue before, but on my local copy of my branch (which I just pushed), this test passes:

```
      deconstructData1:               OK
      deconstructorData2:             OK
      deconstructData2:               OK
      deconstructData3:               OK
    Data
```

michaelpj commented:
> I can't seem to reproduce this failure locally

I agree it's weird, but I suspect you can just revert the change contained in your PR, I suspect it will accept the old one.

thealmarty commented:
That's a static test I don't see why it would do that either. Anyway lmk if you think the parser is suspicious. It seems my fix works.

kozross commented:
@thealmarty - your fix looks like it worked. I also have no idea why this test fails.

michaelpj commented:
What a saga. Thanks!
